docs: Add documentation for Appsets in any namespace (#1249)
argoproj-labs/argocd-operator: docs/usage/appsets-in-any-namespace.md

# ApplicationSets in Any Namespace

**Current feature state**: Beta

!!! note
    This feature is considered beta feature in upstream Argo CD as of now. Some of the implementation details may change over the course of time until it is promoted to a stable status.

> Argo CD supports managing `ApplicationSet` resources in namespaces other than the control plane's namespace. Argo CD administrators can define a certain set of namespaces where `ApplicationSet` resources may be created, updated and reconciled in. 

Reviewer: @svghadi We need a bit more context/background info/concept infor for downstream along with purposes/benefits stating why this features is important and how it helps the users.

In order to manage `ApplicationSet` outside the Argo CD's control plane namespace, two prerequisites must be satisfied:

> 1. The Argo CD should be cluster-scoped

Reviewer: @svghadi In that case, please change this to:

```suggestion
1. The Argo CD instance should be cluster-scoped
```

> 2. The enabled namespace should be entirely covered by the [Apps in Any Namespace](./apps-in-any-namespace.md)

Author: Yes. Updated the 2nd point wth below
> 2. [Apps in Any Namespace](./apps-in-any-namespace.md) should be enabled on target namespaces

## Enable ApplicationSets in a namespace

To enable this feature in a namespace, add the namespace name under `.spec.applicationSet.sourceNamespaces` field in ArgoCD CR.

```yaml
apiVersion: argoproj.io/v1alpha1
kind: ArgoCD
metadata:
  name: example-argocd
spec:
  applicationSet:

>     sourceNamespaces:

Author: Added below sentence at the top of the yaml block to explain what it does
> For example, following configuration will allow `example` Argo CD instance to create & manage `ApplicationSet` resource in `foo` namespace.

      - some-namespace
```

As of now, wildcards are not supported in `.spec.applicationSet.sourceNamespaces`. 

!!! important 

>     Ensure that [Apps in Any Namespace](./apps-in-any-namespace.md) is enabled on target namespace i.e the target namespace name is covered under `.spec.sourceNamespaces` in ArgoCD CR.

Author: Updated with below
> !!! important 
    Ensure that [Apps in Any Namespace](./apps-in-any-namespace.md) is enabled on target namespace i.e the target namespace name is part of `.spec.sourceNamespaces` field in ArgoCD CR.

> The operator creates/modifies below RBAC resources when ApplicationSets in Any Namespace is enabled

Author: It's general

|Name|Kind|Purpose|
|:-|:-|:-|

> |`<argoCDName-argoCDNamespace>-argocd-applicationset-controller`|ClusteRole & ClusterRoleBinding|for `applicationset-controller` to watch & list `ApplicationSets` at cluster-level|

Reviewer: @svghadi what does `&` mean here? Is it both kind together at the same time?

Author: Yes

> |`<argoCDName-argoCDNamespace>-argocd-applicationset-controller`|ClusteRole & ClusterRoleBinding|for `applicationset-controller` to watch & list `ApplicationSets` at cluster-level|

Reviewer: @svghadi Do you mean this - For `ApplicationSet` controller to watch and list `ApplicationSet` resources at cluster-level?

> |`<argoCDName-argoCDNamespace>-applicationset`|Role & RoleBinding|in target namespace for `applicationset-controller` to manage `ApplicationSet` resources|

Reviewer: @svghadi Do you mean this - For `ApplicationSet` controller to manage `ApplicationSet` resources in target namespace?

> |`<argoCDName-argoCDNamespace>-applicationset`|Role & RoleBinding|in target namespace for `applicationset-controller` to manage `ApplicationSet` resources|

Reviewer: @svghadi Please update this content as per the suggestion provided. I see the same sentence and no changes here.

> |`<argoCDName-targetNamespace>`|Role & RoleBinding|in target namespace for `argocd-server` to manipulate `ApplicationSet` resources via UI, API & CLI|

Reviewer: @svghadi what does & mean here? Is it both kind together at the same time?

Author: Yes

Reviewer: @svghadi Please update this content as per the suggestion provided. I see the same sentence and no changes here.

> |`<argoCDName-targetNamespace>`|Role & RoleBinding|in target namespace for `argocd-server` to manipulate `ApplicationSet` resources via UI, API & CLI|

Reviewer: @svghadi Do you mean this - For Argo CD-server to manipulate `ApplicationSet` resources in target namespace through UI, API, or CLI?

- Is it UI, API & CLI?
- UI, API, or CLI?

Also, @svghadi What exactly do you mean by manipulating here? Is it editing/modifying the resources? Please clarify and if need be rewrite with clear wording for better clarity and word usage.

Author: > What exactly do you mean by manipulating here? Is it editing/modifying the resources?
Yes. Creating/modifying/deleteing. Changed `manipulate` to `manage`.

Reviewer: @svghadi Please update this content as per the suggestion provided. I see the same sentence and no changes here.

Additionally, it adds `argocd.argoproj.io/applicationset-managed-by-cluster-argocd` label to the target namespace.

Note that generated `Application` can create resources in any namespace. However, the `Application` itself will be in same namespace as `ApplicationSet`.

## Allow SCM Providers

> By default, whenever ApplicationSets in Any Namespace is enabled, operator disables SCM & PR generators of security reasons. Read upstream [documentation](https://argo-cd.readthedocs.io/en/stable/operator-manual/applicationset/Appset-Any-Namespace/#scm-providers-secrets-consideration) for more details. 

Reviewer: ```suggestion
By default, whenever you enable the ApplicationSets in Any Namespace feature, the Operator disables SCM & PR generators for security reasons. Read upstream [documentation](https://argo-cd.readthedocs.io/en/stable/operator-manual/applicationset/Appset-Any-Namespace/#scm-providers-secrets-consideration) for more details. 
```

Reviewer: @svghadi expand SCM please.

Reviewer: @svghadi Does the word PR generators mean Pull Request generators?

Author: Source Code Manager (SCM) Provider Generator
Pull Request Generator

> To use SCM & PR generators, Argo CD administrators need to explicitly define allowed SCM providers whitelist using `.spec.applicationSet.scmProviders` field in ArgoCD CR. 

Reviewer: @svghadi Please refer to: https://redhat-documentation.github.io/supplementary-style-guide/#_blacklist_and_whitelist

Please use words such as allowlist/passlist instead.

> To use SCM & PR generators, Argo CD administrators need to explicitly define allowed SCM providers whitelist using `.spec.applicationSet.scmProviders` field in ArgoCD CR. 

Reviewer: ```suggestion
To use SCM & PR generators, Argo CD administrators must explicitly define and allow a list of SCM providers using the `.spec.applicationSet.scmProviders` field in the ArgoCD CR. 
```

```yaml
apiVersion: argoproj.io/v1alpha1
kind: ArgoCD
metadata:
  name: example-argocd
spec:
  applicationSet:
    sourceNamespaces:
      - some-namespace

>     scmProviders:

Reviewer: @svghadi Should we provide this content in downstream docs? Please confirm.

If yes, then Should we provide this content as well in downstream docs - https://argo-cd.readthedocs.io/en/stable/operator-manual/applicationset/Appset-Any-Namespace/#scm-providers-secrets-consideration? Please confirm.

      - https://git.mydomain.com/
      - https://gitlab.mydomain.com/
```

> This will configure `applicationset-controller` to allow SCM & PR generators for whitelisted URLs. If any other url is used, it will be rejected by the `applicationset-controller`.

Reviewer: ```suggestion
This will configure `applicationset-controller` to allow the defined URLs for SCM & PR generators. If any other url is used, it will be rejected by the `applicationset-controller`.
```

!!! important

>     Please read upstream [documentation](https://argo-cd.readthedocs.io/en/stable/operator-manual/applicationset/Appset-Any-Namespace/#scm-providers-secrets-consideration) carefully. Misconfiguration could lead to potential security issues.

Reviewer: @svghadi In that case, what all other section from this upstream docs must go into downstream docs related to Applications in any namespace?? Need clarity.

### Things to consider

> - No namespace can be managed by multiple argo-cd instances (cluster scoped or namespace scoped) i.e, only one of either `managed-by` or `applicationset-managed-by-cluster-argocd` labels can be applied to a given namespace. We will be prioritizing `managed-by` label in case of a conflict as this feature is currently in beta, so the new roles/rolebindings will not be created if namespace is already labelled with `managed-by` label, and they will be deleted if a namespace is first added to the `.spec.applicationSet.sourceNamespacs` list and is later also labelled with `managed-by` label.

Reviewer: @svghadi How much of this is downstream-specific? Can we put this as a prerequisite?

Author: It is not a prerequisite but a note on behaviour incase of conflict. No user action is required.

> 

Reviewer: @svghadi Just curious why the following content got removed:

Users will not be create/manage apps and create app resources in the same namespace that is added to sourceNamespaces (as they both require their own labels) out of the box. As a workaround users will have to create custom roles to be able to create app resources in the namespace added to sourceNamespaces.

Author: It is not applicable directly for appset in any namespace feature.

To use SCM Provider & PR generators, Argo CD administrators must explicitly define a list of allowed SCM providers using the `.spec.applicationSet.scmProviders` field in the ArgoCD CR. 

```yaml
apiVersion: argoproj.io/v1beta1
kind: ArgoCD
metadata:
  name: example
spec:
  applicationSet:
    sourceNamespaces:
      - foo
    scmProviders:
      - https://git.mydomain.com/
      - https://gitlab.mydomain.com/
```

This will configure ApplicationSet controller to allow the defined URLs for SCM Provider & PR generators. If any other url is used, it will be rejected by the ApplicationSet controller.

!!! important
    Please read upstream [documentation](https://argo-cd.readthedocs.io/en/stable/operator-manual/applicationset/Appset-Any-Namespace/#scm-providers-secrets-consideration) carefully. Misconfiguration could lead to potential security issues.

### Things to consider

> - No namespace can be managed by multiple argo-cd instances (cluster scoped or namespace scoped) i.e, only one of either `managed-by` or `applicationset-managed-by-cluster-argocd` labels can be applied to a given namespace. We will be prioritizing `managed-by` label in case of a conflict as this feature is currently in beta, so the new roles/rolebindings will not be created if namespace is already labelled with `managed-by` label, and they will be deleted if a namespace is first added to the `.spec.applicationSet.sourceNamespacs` list and is later also labelled with `managed-by` label.

Reviewer 2: This statement is confusing 
> No namespace can be managed by multiple argo-cd instances (cluster scoped or namespace scoped) 

May be we can just remove that line and start from here.
```
only one of either `managed-by` or `applicationset-managed-by-cluster-argocd` labels can be applied to a given namespace. We will be prioritizing `managed-by` label in case of a conflict as this feature is currently in beta, so the new roles/rolebindings will not be created if namespace is already labelled with `managed-by` label, and they will be deleted if a namespace is first added to the `.spec.applicationSet.sourceNamespacs` list and is later also labelled with `managed-by` label.
```

Author: Incorporated your suggestion. Thanks
